### PR TITLE
[5.1] Make controller protected methods unavailable from routes

### PR DESCRIPTION
**Pull request by akalongman**
This PR solves #15783 bug

### CODE REVIEW

**Comment by themsaid** [comment]
I think this is a breaking change, imo I don't see this as a bug and changing the current behaviour might break applications.

**Comment by akalongman** [comment]
Imho possibility to access protected controller methods from routing is potentially security issue

**Comment by Xethron** [comment]
I always thought protected methods are "protected", this is the first time I realize that they could be used in a router. If it does break BC, its probably a good thing, as it will inform developers of problems with their routes/controllers.

**Comment by e1himself** [comment]
@akalongman I think it's better to move this sanity check to `\Illuminate\Routing\Controller::callAction()` method itself.

It's `callAction()` method that has this vulnerability. So it should be fixed to not allow it anymore.

Also you're breaking encapsulation here by assuming it will call the same method with no additional logic.
Imagine Symfony-like controller base class that will call `$this->{$method}Action()` instead.

-----
Anyway, great find @akalongman :+1:

**Comment by akalongman** [comment]
@e1himself I've thinked about implementing this check in `\Illuminate\Routing\Controller::callAction()` but decided to not use Reflection and declined this idea. Using Reflection is heavy operation and I don't think its a good idea.

About encapsulation, I am not sure that is good idea encapsulate controller methods which is used for routes.

**Comment by e1himself** [comment]
@akalongman 
It's just once per request. It costs nothing. 

If choosing between Code Quality and 5 nanosecond always choose Code.

Check Illuminate/Container/Container. There's a lot of Reflection calls performed ~100 times per request. Is it heavy?

Also relevant StackOverflow discusssion: http://stackoverflow.com/a/294584/1895069

**Comment by taylorotwell** [comment]
But the only way for you to route to a protected method is to explicitly declare that you *want* to route to that method in your routes file?

File: src/Illuminate/Routing/ControllerDispatcher.php [hunk 1]
```diff
@@ -154,9 +155,15 @@ public function methodExcludedByOptions($method, array $options)
      * @param  \Illuminate\Routing\Route  $route
      * @param  string  $method
      * @return mixed
+     *
+     * @throws \BadMethodCallException
      */
     protected function call($instance, $route, $method)
     {
+        if (! in_array($method, get_class_methods($instance))) {
```
Comment (taylorotwell) [inline src/Illuminate/Routing/ControllerDispatcher.php]: What if they want it to fall through to __call?

**Comment by ntzm** [comment]
This isn't really a vulnerability, because you have to explicitly tell the router to route to that method, right?

**Comment by akalongman** [comment]
@taylorotwell I can move this sanity check in `Illuminate\Routing\Controller::callAction()`

**Comment by taylorotwell** [comment]
No plans on changing this right now. Laravel will route to the method you tell it to route to in your routes file. You have to explicitly route to this method for Laravel to route to it.

**Pull request closed without merging**